src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
-        ...lists,
-        [list.id]: [list.name, list.value]
+
+      data[proccode] = { args, warp };
+    }
+
+    return (proccode: string): CustomBlockData => {
+      return data[proccode];
+    };
+  }
+
+  interface SerializeTargetOptions {
+    stage: Stage;
+
+    getBroadcastId: GetBroadcastId;
+
+    broadcasts: sb3.Sprite["broadcasts"];
+  }
+
+  function serializeTarget(target: Target, options: SerializeTargetOptions): sb3.Target {
+    // Serialize a target. This function typically isn't used on its own, in
+    // favor of the specialized functions for sprites and stage. It contains
+    // the base code shared across all targets - sounds and costumes, variables
+    // and lists, and, of course, blocks, for example.
+    //
+    // In Scratch 3.0, the representation for the code in a sprite is a flat,
+    // one-dimensional mapping of block ID to block data. To identify which
+    // blocks are the first block in a "script", a topLevel flag is used.
+    // This differs considerably from 2.0, where the scripts property of any
+    // target contained an AST (abstract syntax tree) representation.
+    //
+    // Since all blocks are stored in a flat mapping (and none share an ID),
+    // sb-edit makes use of a "BlockData" system, which is really just a
+    // syntactical wrapper around Object.assign-ing data onto an object.
+    // When a block is serialized, a BlockData store is returned, and this is
+    // combined into the BlockData of whatever is consuming the serialized
+    // data. Eventually, all blocks (and their subblocks, inputs, etc) have
+    // been serialized, and the collected data is stored on the target.
+    //
+    // serializeTarget also handles converting costumes, sounds, variables,
+    // etc into the structures Scratch 3.0 expects.
+
+    const mapToIdObject = (
+      values: Array<{ id: string; [propName: string]: any }>,
+      fn: (x: any) => any
+    ): { [key: string]: any } => {
+      const ret = {};
+      for (const object of values) {
+        ret[object.id] = fn(object);
+      }
+      return ret;
+    };
+
+    const { broadcasts, getBroadcastId, stage } = options;
+
+    const blockData = newBlockData();
+
+    const getCustomBlockData = collectCustomBlockData(target);
+
+    for (const script of target.scripts) {
+      applyBlockData(
+        blockData,
+        serializeBlock(script.blocks[0], {
+          stage,
+          target,
+          getBroadcastId,
+          getCustomBlockData,
+          siblingBlocks: script.blocks,
+          x: script.x,
+          y: script.y
+        }).blockData
+      );
+    }
+
+    const blocks = blockData;
+
+    return {
+      name: target.name,
+      isStage: target.isStage,
+
+      currentCostume: target.costumeNumber,
+      layerOrder: target.layerOrder,
+      volume: target.volume,
+
+      blocks,
+      broadcasts,
+
+      // sb-edit doesn't support comments (as of feb 12, 2020)
+      comments: {},
```

Comment:
nit: add a `TODO` so it's easier to search for? (plus eslint can optionally warn on `TODO`s)